flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/trait/ModifyKindSetTrait.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.`trait`
+
+import org.apache.calcite.plan.{RelOptPlanner, RelTrait, RelTraitDef}
+
+import scala.collection.JavaConversions._
+
+/**
+ * ModifyKindSetTrait is used to describe what modify operation will be produced by this node.
+ */
+class ModifyKindSetTrait(val modifyKindSet: ModifyKindSet) extends RelTrait {
+
+  override def satisfies(relTrait: RelTrait): Boolean = relTrait match {
+    case other: ModifyKindSetTrait =>
+      // it’s satisfied when modify kinds are included in the required set,
+      // e.g. [I,U] satisfy [I,U,D]
+      //      [I,U,D] not satisfy [I,D]
+      this.modifyKindSet.getContainedKinds.forall(other.modifyKindSet.contains)
+    case other: UpdateKindTrait =>
```

Comment:
why `ModifyKindSetTrait` will be compared with `UpdateKindTrait`

Reply:
You are right. This shouldn't be here. It is used to check whether the given UpdateKindTrait is valid. I move this logic into the validation in `FlinkChangelogModeInferenceProgram`.